IntegrationTests/UITests/UITest.cs

```diff
@@ -27,32 +30,47 @@ void HandleApplicationIdle(object sender, EventArgs e)
         }
 
         public static void RunForm<T>(
-            Action showDialog,
-            Func<T, Task> runAsync)
+            Action showForm,
+            Func<T, Task> runTestAsync)
             where T : Form
         {
-            // Avoid using ThreadHelper.JoinableTaskFactory for the outermost operation because we don't want the task
-            // tracked by its collection. Otherwise, test code would not be able to wait for pending operations to
-            // complete.
-            var test = ThreadHelper.JoinableTaskContext.Factory.RunAsync(async () =>
+            // If form.Dipose was called by the async test task, closing the window would be done in a strange order.
+            T form = null;
+            try
             {
-                await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
-                await WaitForIdleAsync();
-                var dialog = Application.OpenForms.OfType<T>().Single();
-                try
-                {
-                    await runAsync(dialog);
-                }
-                finally
+                // Start runTestAsync before calling showForm.
+                // The latter might block until the form is closed, especially if using Application.Run(form).
+
+                // Avoid using ThreadHelper.JoinableTaskFactory for the outermost operation because we don't want the task
+                // tracked by its collection. Otherwise, test code would not be able to wait for pending operations to
+                // complete.
+                var test = ThreadHelper.JoinableTaskContext.Factory.RunAsync(async () =>
                 {
-                    dialog.Close();
-                }
-            });
+                    await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
+                    await WaitForIdleAsync();
+                    form = Application.OpenForms.OfType<T>().Single();
+                    try
+                    {
+                        Assert.True(form.Visible, $"{form.GetType().FullName} {form.Name} should be visible");
+                        await runTestAsync(form);
+                    }
+                    finally
+                    {
+                        form.Close();
+                    }
+                });
 
-            showDialog();
+                showForm();
 
-            // Join the asynchronous test operation so any exceptions are rethrown on this thread
-            test.Join();
+                // Join the asynchronous test operation so any exceptions are rethrown on this thread.
+                using var cts = new CancellationTokenSource(AsyncTestHelper.UnexpectedTimeout);
+                test.Join(cts.Token);
+            }
+            finally
+            {
+                form?.Dispose();
```

Comment:
💭 Can we eliminate the `try`/`finally` above, and just call `Close` here?

```suggestion
                form?.Close();
                form?.Dispose();
```

Reply:
You've lost me. You may find `try finally` ugly. But why should we not dispose the form on exception? If the test passes, `form` will have already been disposed here.

Inserted `form?.Close`.

Reply:
> Inserted form?.Close.

The other one can be removed now.

> If the test passes, form will have already been disposed here.

Where is the other call to dispose?

Reply:
> You've lost me. You may find try finally ugly.

It's not that I find try/finally ugly, it's that I prefer to do things in one location where possible.

> But why should we not dispose the form on exception?

If `Close()` throws an exception, something already went wrong and the test is going to fail. Skipping the call to `Dispose()` in this case doesn't seem problematic.

Reply:
Should just be a using block. 

```c#
using (var x as new WhateverForm())
{
    x......;

}

```

Using is a try finally call.

Reply:
@vbjay, since `showForm` can block, e.g. due to `Application.RunForm(new FormBrowse())`, the Form cannot be created first (also see the comment I have added to `RunForm`).

`using` makes a variable read-only. The following is not allowed:
```
{
    using Form f = null;
    ...
    f = new Form();
    ...
} // f.Dispose is called
```

Reply:
1. `Form.Close` (or `Form.ShowDialog`) seems to call `Dispose`.
  I agree that we should not rely on this implementation detail and call `Dispose` explicitly.

2.
>> But why should we not dispose the form on exception?

> If `Close()` throws an exception, something already went wrong and the test is going to fail. Skipping the call to `Dispose()` in this case doesn't seem problematic.

Maybe yes, if one test fails, we won't accept it. 
But if something goes wrong, we should try to cleanup as much as possible in order to not affect subsequent tests. I don't like to get lots of false negatives.

3. In theory, `showForm` could throw after it has successfully created the Form. That's why `try finally`.